File: CSharpRag/CSharpRagApp/Services/ConfigurationService.cs
using Microsoft.Extensions.Configuration;

namespace CSharpRagApp.Services;

/// <summary>
/// Service for handling configuration loading and validation
/// </summary>
public interface IConfigurationService
{
    RagConfiguration LoadConfiguration();
    (bool IsValid, List<string> MissingSettings) ValidateConfiguration(RagConfiguration config);
}

/// <summary>
/// RAG application configuration
/// </summary>
public class RagConfiguration
{
    public string AzureOpenAIEndpoint { get; set; } = string.Empty;
    public string ChatDeployment { get; set; } = "gpt-4";
    public string SearchEndpoint { get; set; } = string.Empty;
    public string SearchIndex { get; set; } = string.Empty;
    public string SearchApiKey { get; set; } = string.Empty;
}

public class ConfigurationService : IConfigurationService
{
    private readonly IConfiguration _configuration;

    public ConfigurationService(IConfiguration configuration)
    {
        _configuration = configuration;
    }

    public RagConfiguration LoadConfiguration()
    {
        return new RagConfiguration
        {
            AzureOpenAIEndpoint = GetConfigValue("AzureOpenAI:Endpoint", "AZURE_OPENAI_ENDPOINT"),
            ChatDeployment = _configuration["AzureOpenAI:ChatDeployment"] ?? "gpt-4",
            SearchEndpoint = GetConfigValue("AzureSearch:Endpoint", "AZURE_SEARCH_ENDPOINT"),
            SearchIndex = GetConfigValue("AzureSearch:IndexName", "AZURE_SEARCH_INDEX_NAME"),
            SearchApiKey = GetConfigValue("AzureSearch:ApiKey", "AZURE_SEARCH_API_KEY")
        };
    }

    public (bool IsValid, List<string> MissingSettings) ValidateConfiguration(RagConfiguration config)
    {
        var missingSettings = new List<string>();

        if (string.IsNullOrEmpty(config.AzureOpenAIEndpoint))
            missingSettings.Add("AZURE_OPENAI_ENDPOINT");

        if (string.IsNullOrEmpty(config.SearchEndpoint))
            missingSettings.Add("AZURE_SEARCH_ENDPOINT");

        if (string.IsNullOrEmpty(config.SearchIndex))
            missingSettings.Add("AZURE_SEARCH_INDEX_NAME");

        return (missingSettings.Count == 0, missingSettings);
    }

    private string GetConfigValue(string configKey, string envVarKey)
    {
        return _configuration[configKey] 
            ?? Environment.GetEnvironmentVariable(envVarKey) 
            ?? string.Empty;
    }
}

File: PythonRag/python_rag_app/services/configuration_service.py
"""Configuration service for RAG application."""

import os
from dataclasses import dataclass
from typing import Protocol, Tuple


@dataclass
class RagConfiguration:
    """RAG application configuration."""

    azure_openai_endpoint: str = ""
    chat_deployment: str = "gpt-4"
    search_endpoint: str = ""
    search_index: str = ""
    search_api_key: str = ""


class IConfigurationService(Protocol):
    """Interface for configuration service."""

    def load_configuration(self) -> RagConfiguration:
        """Load configuration from environment."""
        ...

    def validate_configuration(
        self, config: RagConfiguration
    ) -> Tuple[bool, list[str]]:
        """Validate configuration and return (is_valid, missing_settings)."""
        ...


class ConfigurationService:
    """Service for handling configuration loading and validation."""

    def load_configuration(self) -> RagConfiguration:
        """Load configuration from environment variables.

        Returns:
            RagConfiguration with loaded settings.
        """
        return RagConfiguration(
            azure_openai_endpoint=os.environ.get("AZURE_OPENAI_ENDPOINT", ""),
            chat_deployment=os.environ.get(
                "AZURE_OPENAI_CHAT_DEPLOYMENT", "gpt-4"
            ),
            search_endpoint=os.environ.get("AZURE_SEARCH_ENDPOINT", ""),
            search_index=os.environ.get("AZURE_SEARCH_INDEX_NAME", ""),
            search_api_key=os.environ.get("AZURE_SEARCH_API_KEY", ""),
        )

    def validate_configuration(
        self, config: RagConfiguration
    ) -> Tuple[bool, list[str]]:
        """Validate configuration and return missing settings.

        Args:
            config: Configuration to validate.

        Returns:
            Tuple of (is_valid, list of missing setting names).
        """
        missing_settings = []

        if not config.azure_openai_endpoint:
            missing_settings.append("AZURE_OPENAI_ENDPOINT")

        if not config.search_endpoint:
            missing_settings.append("AZURE_SEARCH_ENDPOINT")

        if not config.search_index:
            missing_settings.append("AZURE_SEARCH_INDEX_NAME")

        return (len(missing_settings) == 0, missing_settings)
